### PR TITLE
8253424: Add support for running pre-submit testing using GitHub Actions

### PR DESCRIPTION
**Pull request by rwestberg**
A few days ago I posted an initial version of the necessary configuration required to run pre-submit build and tests for JDK main-line contributions using GitHub Actions [2] and the free tier [3] available to everyone working with open source repositories. I've incorporated the feedback into an updated version that I believe is ready to be integrated.

If this is integrated into the `master` branch, future branches created and updated in personal forks will build and run the basic tier 1 tests as described in this configuration, on Linux, Windows and macOS (all on x64). It's of course possible for any contributor to opt out fully or partially of these automatic runs in a few different ways.

To opt out completely, a contributor can simply disable GitHub Actions on their personal fork, and no further jobs will be executed. Another option is to add a repository secret [4] with the name `JDK_SUBMIT_FILTER` set to any value. If this is set, only branches prefixed with `submit/` will be subject to automatic build and test. This can also be further refined by adding a repository secret named `JDK_SUBMIT_PLATFORMS` with a value such as `Linux x64, Windows x64` to limit automatic build and test to these two platforms. It will still be possible to run the tests on any branch and/or platform by manually triggering the workflow.

To see what this looks like in practice, an example run can be found here: https://github.com/rwestberg/jdk/actions/runs/265131985 (note that there is currently a failing test on Windows which is tracked by JDK-8249095, which should probably be resolved before this change is integrated).

Best regards,
Robin

[1] https://mail.openjdk.java.net/pipermail/jdk-dev/2020-September/004736.html
[2] https://github.com/features/actions
[3] https://docs.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits
[4] https://docs.github.com/en/actions/reference/encrypted-secrets

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253424](https://bugs.openjdk.java.net/browse/JDK-8253424): Add support for running pre-submit testing using GitHub Actions

### Reviewers
 * [Erik Helin](https://openjdk.java.net/census#ehelin) (@edvbld - **Reviewer**) ⚠️ Review applies to 441a18b7fa75635d7b06ed203018ea9e89af74fe
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to 441a18b7fa75635d7b06ed203018ea9e89af74fe

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/284/head:pull/284`
`$ git checkout pull/284`

### CODE REVIEW

**Comment by erikj79** [comment]
Is there any way to derive the version strings for dependencies from a different source? Once pushed, this file will require updating whenever the bootjdk, jtreg version or gtest version is updated.

**Comment by rwestberg** [comment]
Certainly, the prerequisites step can checkout additional repositories and run shell commands to extract variables to be used for later steps. Do we have any suitable source for these versions? We'll also need download links for the different flavors of the bootjdk (and ideally the sha256 hash, although the verification step could perhaps be skipped).

**Comment by erikj79** [comment]
We don't have a good source currently no, but something to keep in mind. Ideally we would want to create a common source for both jib-profiles.js and the gitactions to read from. I would prefer something that can be read both as a properties file and sourced as a shell script, much like the version-numbers file currently is. That could be done as a separate change though.

**Comment by rwestberg** [comment]
Sure, should not be that hard to parse something similar. The GitHub actions will probably need it in JSON format at some point, but nothing a little `sed -e '1i {' -e 's/#.*//g' -e 's/"//g' -e 's/\(.*\)=\(.*\)/"\1": "\2",/g' -e '$s/,\s\{0,\}$/\}/'` won't solve. Any suggestion for the location and name of such a file? I'm thinking it would contain entries like this:

```
BOOT_JDK_VERSION="14.0.2"
JTREG_VERSION="jtreg5.1-b01"
GTEST_VERSION="release-1.8.1"
LINUX_X64_BOOT_JDK_FILENAME="openjdk-14.0.2_linux-x64_bin.tar.gz"
LINUX_X64_BOOT_JDK_URL="https://download.java.net/java/GA/jdk14.0.2/205943a0976c4ed48cb16f1043c5c647/12/GPL/openjdk-14.0.2_linux-x64_bin.tar.gz"
LINUX_X64_BOOT_JDK_SHA256="91310200f072045dc6cef2c8c23e7e6387b37c46e9de49623ce0fa461a24623d"
```

**Comment by erikj79** [comment]
The other primary consumer of this is make/conf/jib-profiles.js. The make/conf dir would make sense to me.

The challenge here is creating a set of variables that are suitable enough for both config files to consume. For BootJDK, we never bothered with bumping the version for updates, and I very much doubt we will do that in the future for github actions, so a plain major version 14, and soon 15, would be preferred from my point of view. This is however not enough for jib-profiles.js (yet) so we can't really share bootjdk config for now anyway. For jtreg, we specify 5.1 and b01 as two separate metadata values. For gtest we specify the version as 1.8.1.

**Comment by rwestberg** [comment]
I added `make/conf/test-dependencies` with version numbers specified on the format that `jib-profiles.js` would expect. Actually using them from that file as well could perhaps be a separate task though. :)

**Comment by erikj79** [comment]
The version-numbers file (which is also a shared properties style file) is not using quotes for values, which is fine as long as there are no spaces. I believe if you read it as a properties file, you need to strip the quotes if you have them. I prefer if version-numbers and test-dependencies using the same format.

Otherwise this looks good to me!

**Comment by rwestberg** [comment]
Sounds reasonable, didn't notice that discrepancy! Fixed in the latest commit.

**Comment by rwestberg** [comment]
After some feedback on the jdk-dev mailing list (thanks @jaikiran!) I took a second look at removing artifacts that are only created for passing between the build and test steps. Turns out that there is an API for this (it's still in preview, but seems to work fine) that can be used for this. When this API is finalized we can update that part, but it's not a deal-breaker even if it should change, as the test runs will still complete successfully. With this API available, it's also possible to publish a single artifact containing all test results, regardless of outcome.

**Review by edvbld — APPROVED** [review]
Looks good to me, awesome work @rwestberg!

**Review by erikj79 — APPROVED** [review]
